doc/compiler/osr/OSR.md

```diff
@@ -0,0 +1,237 @@
+
+**Fear point analysis**: is a backward union analysis that determines whether it's possible for execution to reach a fear point without first hitting an OSR yield point. OSR guards are inserted after each OSR yield point at which this is possible. This ensures that every fear point is protected by the closest possible OSR guard. In addition, all of the resulting OSR guards are in positions where it's possible to induce OSR because HCR guard analysis has eliminated any fear points that would require OSR guards where OSR is uninducible.
+
+OSR guards are placed after OSR yield points and behave similarly to HCR guards. The guard will not necessarily be placed immediately after the OSR yield point,
+rather several stores may be located between the two. These will include stores to pending push slots, to determine the stack at the transition point (as these are used for rematerialization),
+and stores of call or allocation results. The fall through path will point to the inlined code,
+but the taken path will point to **OSR induce block**. OSR induce block will contain call to the **jitInduceOSRAtCurrentPC** helper and will have exception edges to the **OSR catch block**.
+
+As an optimization, we might have a store of the call result into an auto other than a pending push temp is that call results are often immediately stored to autos in the bytecode. Since we need to store it to the auto anyway, we want to avoid storing to a pending push temp, then potentially inducing OSR, and then storing again to the original auto. Instead we notice that the call result is stored into an auto (and presumably that what we store into the auto is always the result of the call, i.e. there's no branch target in between), and we push the OSR point forward past the store.
+
+Runtime assumptions are added to patch the inserted OSR guards in the event of a method redefinition,
+triggering the transition of execution to the JVM via the OSR mechanism. The OSR guards are permanently patched during an OSR yield to the JVM at which invalidation happens( e.g. a redefinition occurs). An important distinction is that all OSR guards will be patched in the event of any redefinition,
+rather than just the required set to protect a removed HCR guard. This is due to the high overhead of calculating these sets for individual HCR guards.
+
+## OSR and virtual guard merging
+
+As an optimization, we can mark some existing virtual guard as an OSR guard, therefore merging them together. This is achieved by making virtual guard kill fear. The disadvantage of the guard  merging is that the merged guard cannot be removed anymore, while the original virtual guard could've been removed as a result of later optimizaitons.
```

Comment:
Oh, I mentioned it because we specifically introduced a change where we don't do guard merging for Vector API methods since it resulted in extra virtual guards that could've been eliminated otherwise. Maybe we don't need that change anymore?

Reply:
Ah, no, that change is still needed because the OSR guard would still remain, and it would still target the cold call. I suppose it might possibly be removed later based on OSR guard analysis, but we can't rely on the OSR guard being redundant whenever the virtual guard is guaranteed to pass.

Maybe say that the merged guard cannot be *completely* removed anymore? But even that seems misleading given the possibility that the OSR guard could happen to be redundant. Taking a stab at it:
> The disadvantage of guard merging is that a later type analysis that would be sufficient to remove a virtual guard won't be able to also remove the corresponding OSR guard, which likely prevents the elimination of a control flow diamond.